### PR TITLE
Add reusable kubernetes-backend proxy client

### PR DESCRIPTION
**Pull request by RubenV-dev**
## Hey, I just made a Pull Request!
I am extending the `KubernetesApi` interface to include a proxy method that front-end plugins can use to consume the `KubernetesProxy` endpoint. The new interface can be seen below:
```typescript
export interface KubernetesApi {
  getObjectsByEntity(
    requestBody: KubernetesRequestBody,
  ): Promise<ObjectsByEntityResponse>;
  getClusters(): Promise<
    {
      name: string;
      authProvider: string;
      oidcTokenProvider?: string | undefined;
    }[]
  >;
  getWorkloadsByEntity(
    request: WorkloadsByEntityRequest,
  ): Promise<ObjectsByEntityResponse>;
  getCustomObjectsByEntity(
    request: CustomObjectsByEntityRequest,
  ): Promise<ObjectsByEntityResponse>;
  proxy(options: {
    clusterName: string;
    path: string;
    init?: RequestInit;
  }): Promise<Response>;
}
```
This feature implementation and more context can be found in issue #16065. 

This change allow a front-end plugin to use the proxy endpoint like in the following example:
```typescript
const kubernetesApi = useApi(kubernetesApiRef);
await kubernetesApi.proxy(CLUSTER_NAME, '/api/v1/namespaces');
```
I also backflll missing tests for the `KubernetesBackendClient`

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [X] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by mclarke47** [comment]
What a coincidence I was just making some similar proxy changes but was going to ask how best to proceed 

https://github.com/backstage/backstage/pull/17120

File: plugins/kubernetes/api-report.md
Comment (mclarke47) [inline plugins/kubernetes/api-report.md:268]: What's the `RequestInit` for?

File: plugins/kubernetes/api-report.md
Comment (mclarke47) [inline plugins/kubernetes/api-report.md:265]: I think it would be nice for the proxy to be it's own API that we could add common methods to (like fetching logs etc)?

File: plugins/kubernetes/api-report.md
Comment (jamieklassen) [reply plugins/kubernetes/api-report.md:268]: If you want to use a different HTTP verb, send a request body, add headers, etc. Same reason you'd want it when making a normal `fetch` call. This is discussed in #16065

File: plugins/kubernetes/api-report.md
Comment (mclarke47) [reply plugins/kubernetes/api-report.md:265]: Yeah, this is something we can add later or I can take a first pass at it in my logging PR. I think it makes sense to merge this PR first and me to rebase in that case

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (Rugvip) [inline plugins/kubernetes/src/api/KubernetesBackendClient.ts:160]: I think think you want to use the exact same response handling here, 404 could come back for other reasons than the plugin not being installed. In general I think it might be best to just forward the response directly instead? There could be cases where you for example want to stream out data, you're not dealing with JSON, etc.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice! 😁 

Just a few nits

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (RubenV-dev) [reply plugins/kubernetes/src/api/KubernetesBackendClient.ts:160]: This does make sense, i just tried to reuse some components but those edgecases definitely make me want to change from using the handleResponse function and just returning the response directly. I will rework this.

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (RubenV-dev) [reply plugins/kubernetes/src/api/KubernetesBackendClient.ts:151]: Will look into this

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (jamieklassen) [reply plugins/kubernetes/src/api/KubernetesBackendClient.ts:160]: On this same topic, I noticed the following coverage:
```console
% yarn test --watchAll=false --collectCoverageFrom '**/KubernetesBackendClient.ts' --coverage plugins/kubernetes/src/api/KubernetesBackendClient.test.ts
 PASS   @backstage/plugin-kubernetes  plugins/kubernetes/src/api/KubernetesBackendClient.test.ts
  KubernetesBackendClient
    ✓ hits the /clusters API (50 ms)
    ✓ hits the /resources/custom/query API (5 ms)
    ✓ hits the /services/{entityName} api (4 ms)
    ✓ hits the /resources/workloads/query API (3 ms)
    ✓ hits the /proxy API (5 ms)
    ✓ /proxy API throws a ERROR_NOT_FOUND if the cluster in the request is not found (5 ms)
    ✓ hits /proxy api when signed in as a guest (6 ms)

----------------------------|---------|----------|---------|---------|-------------------
File                        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
----------------------------|---------|----------|---------|---------|-------------------
All files                   |   85.71 |       70 |     100 |   85.36 |
 KubernetesBackendClient.ts |   85.71 |       70 |     100 |   85.36 | 46-56
----------------------------|---------|----------|---------|---------|-------------------
Test Suites: 1 passed, 1 total
Tests:       7 passed, 7 total
Snapshots:   0 total
Time:        3.566 s
Ran all test suites matching /plugins\/kubernetes\/src\/api\/KubernetesBackendClient.test.ts/i
```

and it's striking that lines the uncovered lines are precisely the ones where the response from the kubernetes-backend is not successful:

https://github.com/backstage/backstage/blob/0ebbfd45c62bb05c1780a5d7184a5dae2fa0580a/plugins/kubernetes/src/api/KubernetesBackendClient.ts#L46-L56

Can we add tests to cover the cases when the relevant backend API (including proxy) returns an error status?

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (jamieklassen) [inline plugins/kubernetes/src/api/KubernetesBackendClient.ts:157]: I think what @Rugvip means is
```suggestion
    const headers = {
      ...options.init?.headers,
      [`Backstage-Kubernetes-Cluster`]: options.clusterName,
      [`Backstage-Kubernetes-Authorization`]: `Bearer ${k8sToken}`,
      ...(identityResponse.token && {Authorization: `Bearer ${identityResponse.token}`})
    };
```

File: plugins/kubernetes/src/api/KubernetesBackendClient.test.ts
Comment (RubenV-dev) [reply plugins/kubernetes/src/api/KubernetesBackendClient.test.ts:265]: Sure thing

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (RubenV-dev) [reply plugins/kubernetes/src/api/KubernetesBackendClient.ts:157]: Thanks this helps, i was exploring how that expression gets evaluated because its the first time ive seen it. Pretty neat trick. Refactored conditional.

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (RubenV-dev) [reply plugins/kubernetes/src/api/KubernetesBackendClient.ts:140]: Im wondering if there are any examples where lightweight caching is used in the project. Ive looked at a few example plugins like lighthouse,newrelic,github, and pagerduty but dont seem to quite find what im looking for. Just curious if there is a library that we leverage across the project. Also Im experimenting with adding just a object with the cluster array and a timestamp and im wondering, how much of a difference between each api call should i allow before a new api call is made.

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (RubenV-dev) [reply plugins/kubernetes/src/api/KubernetesBackendClient.ts:160]: yes, working on test coverage for these cases

File: plugins/kubernetes/src/api/KubernetesBackendClient.test.ts
Comment (RubenV-dev) [reply plugins/kubernetes/src/api/KubernetesBackendClient.test.ts]: Done

File: plugins/kubernetes/src/api/KubernetesBackendClient.test.ts
Comment (RubenV-dev) [reply plugins/kubernetes/src/api/KubernetesBackendClient.test.ts]: Done

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (RubenV-dev) [reply plugins/kubernetes/src/api/KubernetesBackendClient.ts]: Done

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (RubenV-dev) [reply plugins/kubernetes/src/api/KubernetesBackendClient.ts]: Done

File: plugins/kubernetes/src/kubernetes-auth-provider/types.ts
Comment (Rugvip) [inline plugins/kubernetes/src/kubernetes-auth-provider/types.ts:24]: Had missed that we needed to add this. I think we should take a bit of time to make sure this is forwards compatible. At minimum I'm feeling we switch up the return value to be `Promise<{ token: string }>` so that we're able to expand it in the future. A name we use for this in many other places is `getCredentials`, with that same return value, do you think that'd be alright too?

**Review by Rugvip — CHANGES_REQUESTED** [review]

File: plugins/kubernetes/src/kubernetes-auth-provider/types.ts
Comment (RubenV-dev) [reply plugins/kubernetes/src/kubernetes-auth-provider/types.ts:24]: Clarifying question: was this suggestion just for the method getBearerToken() in the KubernetesAuthProvider interface? Or should i be renaming the getBearertoken method i introduce to the KubernetesAuthProvidersApi interface as well?

Comment (RubenV-dev) [reply plugins/kubernetes/src/kubernetes-auth-provider/types.ts]: Done

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (Rugvip) [reply plugins/kubernetes/src/api/KubernetesBackendClient.ts:142]: Nothing too fancy, just a `Map<string, { authProvider, timestamp }>` or w/e is needed with a 1-5min timeout. Also avoid caching the credentials. Don't wanna block this PR on adding this, and I'll leave it to @backstage/kubernetes-maintainers whether it's something they think improves this.

File: .changeset/sweet-eels-hunt.md
Comment (Rugvip) [inline .changeset/sweet-eels-hunt.md:2]: This is not a breaking change according to our versioning policy, as it's only breaking API producers, you could move this to `patch`.

**Review by Rugvip — APPROVED** [review]
Thanks! 👍 

Good to go from my point of view, will leave final review and merge to @backstage/kubernetes-maintainers

File: .changeset/sweet-eels-hunt.md
Comment (RubenV-dev) [reply .changeset/sweet-eels-hunt.md:2]: Ah ok will change asap.

Comment (RubenV-dev) [reply .changeset/sweet-eels-hunt.md]: Done

**Comment by mclarke47** [comment]
Since he is more familiar I'll leave this to @jamieklassen

File: plugins/kubernetes/src/api/KubernetesBackendClient.ts
Comment (jamieklassen) [reply plugins/kubernetes/src/api/KubernetesBackendClient.ts:142]: I'll write up an issue for this improvement so we can merge this now!

**Pull request merged**
